src/spartan/batched.rs

```diff
@@ -0,0 +1,645 @@
+    VerifierKey {
+      vk_ee,
+      S: shapes,
+      digest: OnceCell::new(),
+    }
+  }
+}
+
+impl<G: Group, EE: EvaluationEngineTrait<G>> SimpleDigestible for VerifierKey<G, EE> {}
+
+impl<G: Group, EE: EvaluationEngineTrait<G>> DigestHelperTrait<G> for VerifierKey<G, EE> {
+  /// Returns the digest of the verifier's key.
+  fn digest(&self) -> G::Scalar {
+    self
+      .digest
+      .get_or_try_init(|| {
+        let dc = DigestComputer::<G::Scalar, _>::new(self);
+        dc.digest()
+      })
+      .cloned()
+      .expect("Failure to retrieve digest!")
+  }
+}
+
+impl<G: Group, EE: EvaluationEngineTrait<G>> BatchedRelaxedR1CSSNARKTrait<G>
+  for BatchedRelaxedR1CSSNARK<G, EE>
+{
+  type ProverKey = ProverKey<G, EE>;
+
+  type VerifierKey = VerifierKey<G, EE>;
+
+  fn setup(
+    ck: &CommitmentKey<G>,
+    S: &[R1CSShape<G>],
+  ) -> Result<(Self::ProverKey, Self::VerifierKey), NovaError> {
+    let (pk_ee, vk_ee) = EE::setup(ck);
+
+    let S = S.iter().map(|s| s.pad()).collect();
+
+    let vk = VerifierKey::new(S, vk_ee);
+
+    let pk = ProverKey {
+      pk_ee,
+      vk_digest: vk.digest(),
+    };
+
+    Ok((pk, vk))
+  }
+
+  fn prove(
+    ck: &CommitmentKey<G>,
+    pk: &Self::ProverKey,
+    S: &[R1CSShape<G>],
+    U: &[RelaxedR1CSInstance<G>],
+    W: &[RelaxedR1CSWitness<G>],
+  ) -> Result<Self, NovaError> {
+    let num_instances = U.len();
+    // Pad shapes and ensure their sizes are correct
+    let S = S
+      .iter()
+      .map(|s| {
+        let s = s.pad();
+        if s.is_regular_shape() {
+          Ok(s)
+        } else {
+          Err(NovaError::InternalError)
+        }
+      })
+      .collect::<Result<Vec<_>, _>>()?;
+
+    // Pad (W,E) for each instance
+    let W = W
+      .iter()
+      .zip(S.iter())
+      .map(|(w, s)| w.pad(s))
+      .collect::<Vec<RelaxedR1CSWitness<G>>>();
+
+    let mut transcript = G::TE::new(b"BatchedRelaxedR1CSSNARK");
+
+    transcript.absorb(b"vk", &pk.vk_digest);
+    U.iter().for_each(|u| {
+      transcript.absorb(b"U", u);
+    });
+
+    let (polys_W, polys_E): (Vec<_>, Vec<_>) = W.into_iter().map(|w| (w.W, w.E)).unzip();
+
+    // Append public inputs to W: Z = [W, u, X]
+    let polys_Z = polys_W
+      .iter()
+      .zip(U.iter())
+      .map(|(w, u)| [w.clone(), vec![u.u], u.X.clone()].concat())
+      .collect::<Vec<Vec<_>>>();
+
+    let (num_rounds_x, num_rounds_y): (Vec<_>, Vec<_>) = S
+      .iter()
+      .map(|s| (s.num_cons.log_2(), s.num_vars.log_2() + 1))
+      .unzip();
+    let num_rounds_x_max = *num_rounds_x.iter().max().unwrap();
+    let num_rounds_y_max = *num_rounds_y.iter().max().unwrap();
+
+    // Generate tau polynomial corresponding to eq(τ, τ², τ⁴ , …)
+    // for a random challenge τ
+    let tau = transcript.squeeze(b"t")?;
+    let polys_tau = num_rounds_x
+      .iter()
+      .map(|&num_rounds_x| PowPolynomial::new(&tau, num_rounds_x).evals())
+      .map(MultilinearPolynomial::new)
+      .collect::<Vec<_>>();
```

Comment:
I follow this, it is extremely wasteful because you're computing all the prefix powers over in every iterations. It would be better to compute all the powers once, then create a `PowPolynomial` per power — or even just add a `PowPolynomial::new_up_to()` (or whatever).

Reply:
I'll make a PR for this.

Reply:
#148